File: python/find_the_duplicate_number.py
class Solution:
    """
    @param nums: an array containing n + 1 integers which is between 1 and n
    @return: the duplicate one
    """
    def findDuplicate(self, nums):
        # write your code here
        n = len(nums)-1
        lf,rt = 1, n
        while lf <=rt:
            mid = (lf+rt)//2
            cnt =0
            for i in range(n+1):
                if nums[i] <= mid:
                    cnt += 1 
            if cnt > mid:
                rt = mid-1
            else:
                lf = mid +1 
        
        return lf 

test = Solution()
nums = [5,5,4,3,2,1]
res = test.findDuplicate(nums)
print(res)

File: cpp/favourite/FindtheDuplicateNumber.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int res=0;
        for (int i=0; i<32; i++) {
            int cnt1 = 0, cnt2 = 0, bit = 1<<i;
            for (int k=0; k<nums.size(); k++) {
                if (k & bit) cnt1++;
                if (nums[k] & bit) cnt2++;
            }
            if (cnt1 < cnt2) res +=bit;
        }
        return res;
    }
};